File: bundles/io.github.linkedfactory.service/src/test/scala/io/github/linkedfactory/service/test/JsonFormatParserTest.scala
/*
 * Copyright (c) 2022 Fraunhofer IWU.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.github.linkedfactory.service.test

import io.github.linkedfactory.core.kvin.{Kvin, KvinTuple, Record}
import io.github.linkedfactory.service.util.JsonFormatParser
import net.enilink.komma.core.URIs
import net.liftweb.common.{Failure, Full}
import net.liftweb.json.JsonAST._
import net.liftweb.json.JsonDSL._
import org.junit.{Assert, Test}
import org.scalatest.Assertions.fail


class JsonFormatParserTest {

  @Test
  def test(): Unit = {
    val context = Kvin.DEFAULT_CONTEXT
    val time = System.currentTimeMillis
    val root = URIs.createURI("http://example.root")
    val simpleJson: JValue =
      ("@context" -> ("pref" -> "http://test1.example/")) ~
        ("pref" -> ("pref:rest" -> "val") ~ ("pref2:pref3" -> "val2"))
    var tuple = JsonFormatParser.parseItem(root, context, simpleJson, time).head.head
    Assert.assertEquals("http://test1.example/", tuple.item.toString)
    Assert.assertEquals("http://test1.example/rest", tuple.property.toString)
    Assert.assertEquals("val", tuple.value)
    tuple = JsonFormatParser.parseItem(root, context, simpleJson, time).head.tail.head
    Assert.assertEquals("http://test1.example/", tuple.item.toString)
    Assert.assertEquals("pref2:pref3", tuple.property.toString)
    Assert.assertEquals("val2", tuple.value)

    val withoutContext = ("pref" -> ("pref:rest" -> "val") ~ ("pref2:pref3" -> "val2"))
    tuple = JsonFormatParser.parseItem(root, context, withoutContext, time).head.head
    Assert.assertEquals("http://example.root/pref", tuple.item.toString)
    Assert.assertEquals("pref:rest", tuple.property.toString)
    Assert.assertEquals("val", tuple.value)

    val withMultiContext = ("@context" -> ("pref" -> "http://test1.example/") ~ ("pref2" -> "http://pref2.example/")) ~
      ("@context" -> ("pref" -> "http://test2.example/")) ~
      ("pref" -> ("pref:rest" -> "val") ~ ("pref2:rest" -> "val2"))
    tuple = JsonFormatParser.parseItem(root, context, withMultiContext, time).head.head
    Assert.assertEquals("http://test2.example/", tuple.item.toString)
    Assert.assertEquals("http://test2.example/rest", tuple.property.toString)
    Assert.assertEquals("val", tuple.value)
    tuple = JsonFormatParser.parseItem(root, context, withMultiContext, time).head.tail.head
    Assert.assertEquals("http://test2.example/", tuple.item.toString)
    Assert.assertEquals("http://pref2.example/rest", tuple.property.toString)
    Assert.assertEquals("val2", tuple.value)

    val prefixInContext = ("@context" -> ("pref" -> "http://test1.example/") ~ ("pref2" -> "pref:pref1")) ~
      ("@context" -> ("pref" -> "http://test2.example/")) ~
      ("pref" -> ("pref:rest" -> "val") ~ ("pref2" -> "val2"))
    tuple = JsonFormatParser.parseItem(root, context, prefixInContext, time).head.tail.head
    Assert.assertEquals(tuple.property.toString, "http://test1.example/pref1")

    val multiPrefixes = ("@context" -> ("pref" -> "http://test1.example/") ~ ("pref2" -> "pref:pref1")) ~
      ("pref:pref1/pref2" -> ("pref:rest" -> "val") ~ ("pref2" -> "val2"))
    tuple = JsonFormatParser.parseItem(root, context, multiPrefixes, time).head.head
    Assert.assertEquals(tuple.item.toString, "http://test1.example/pref1/pref2")
  }

  @Test
  def testNested(): Unit = {
    val context = Kvin.DEFAULT_CONTEXT
    val time = System.currentTimeMillis
    val root = URIs.createURI("http://example.root")
    val nested = "item" -> ("p1" -> "v1") ~
      ("nested" -> ("value", ("p1" -> "v1") ~ ("p2" -> ("p3", "v3") ~ ("p4", "v4"))))
    val parsed = JsonFormatParser.parseItem(root, context, nested, time)
    val expected = Full(List(
      new KvinTuple(URIs.createURI("http://example.root/item"),
        URIs.createURI("http://example.root/p1"), Kvin.DEFAULT_CONTEXT, time, "v1"),
      new KvinTuple(URIs.createURI("http://example.root/item"),
        URIs.createURI("http://example.root/nested"), Kvin.DEFAULT_CONTEXT, time,
          new Record(URIs.createURI("http://example.root/p1"), "v1").append(
            new Record(URIs.createURI("http://example.root/p2"),
              new Record(URIs.createURI("http://example.root/p3"), "v3").append(
                new Record(URIs.createURI("http://example.root/p4"), "v4")
              )
            )
          )
      )
    ))
    Assert.assertEquals(expected, parsed)
  }

  @Test
  def shouldFail(): Unit = {
    val context = Kvin.DEFAULT_CONTEXT
    val time = System.currentTimeMillis
    val root = URIs.createURI("http://example.root")
    val itemError: JValue =
      ("some item" -> ("http://example.org/property" -> "val"))
    var result = JsonFormatParser.parseItem(root, context, itemError, time)
    assert(result.isInstanceOf[Failure])
    result match {
      case Failure(msg, _, _) =>
        assert(msg.contains("Invalid URI: some item"))
      case _ => fail("Expected Invalid URI: some item, got " + result)
    }

    val propertyError = ("item" -> ("http://example.org/my property" -> "val"))
    result = JsonFormatParser.parseItem(root, context, propertyError, time)
    assert(result.isInstanceOf[Failure])
    result match {
      case Failure(msg, _, _) =>
        assert(msg.contains("Invalid URI: http://example.org/my property"))
      case _ => fail("Expected Invalid URI: http://example.org/my property, got " + result)
    }
  }
}

File: bundles/io.github.linkedfactory.core/src/test/java/io/github/linkedfactory/core/kvin/util/JsonFormatParserTest.java
/*
 * Copyright (c) 2024 Fraunhofer IWU.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.github.linkedfactory.core.kvin.util;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import io.github.linkedfactory.core.kvin.Kvin;
import io.github.linkedfactory.core.kvin.KvinTuple;
import io.github.linkedfactory.core.kvin.Record;
import net.enilink.komma.core.URIs;
import org.junit.Test;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

public class JsonFormatParserTest {
	private final ObjectMapper mapper = new ObjectMapper();

	@Test
	public void shouldParseJsonResource() throws Exception {
		List<KvinTuple> tuples = parse(readResource("/JsonFormatParserTestContent.json"));

		assertEquals(11, tuples.size());
		assertTrue(tuples.get(2).value instanceof java.lang.Integer);
		assertTrue(tuples.get(3).value instanceof java.math.BigInteger);
		assertTrue(tuples.get(4).value instanceof java.lang.Double);
		assertTrue(tuples.get(5).value instanceof java.lang.Long);
		assertTrue(tuples.get(6).value instanceof java.lang.Boolean);
		assertTrue(tuples.get(7).value instanceof Record);
		assertTrue(tuples.get(10).value instanceof Record);
	}

	@Test
	public void shouldResolvePrefixesFromContext() throws Exception {
		String json = """
		{
		  "@context": {"pref": "http://test1.example/", "pref2": "http://pref2.example/"},
		  "@context": {"pref": "http://test2.example/"},
		  "pref": {
		    "pref:rest": [{"value": "val"}],
		    "pref2:rest": [{"value": "val2"}]
		  }
		}""";

		List<KvinTuple> tuples = parse(json, 1619424246100L);

		assertEquals(2, tuples.size());
		assertEquals("http://test2.example/", tuples.getFirst().item.toString());
		assertEquals("http://test2.example/rest", tuples.get(0).property.toString());
		assertEquals("val", tuples.get(0).value);
		assertEquals("http://test2.example/", tuples.get(1).item.toString());
		assertEquals("http://pref2.example/rest", tuples.get(1).property.toString());
		assertEquals("val2", tuples.get(1).value);
	}

	@Test
	public void shouldParseNestedRecords() throws Exception {
		ObjectNode root = mapper.createObjectNode();
		ObjectNode item = root.putObject("http://example.root/item");
		ArrayNode values = item.putArray("http://example.root/nested");
		ObjectNode value = values.addObject().putObject("value");
		value.put("msg", "Error 1");
		value.put("nr", 1);
		value.putObject("test_prop").put("msg", "test");
		value.putObject("id_prop").put("@id", "http://example.org/properties/test3");

		List<KvinTuple> tuples = parse(root, 1619424246100L);
		Record expectedValue = new Record(URIs.createURI("msg"), "Error 1").append(new Record(URIs.createURI("nr"), 1).append(new Record(URIs.createURI("test_prop"), new Record(URIs.createURI("msg"), "test")).append(new Record(URIs.createURI("id_prop"), URIs.createURI("http://example.org/properties/test3")))));
		List<KvinTuple> expected = List.of(new KvinTuple(URIs.createURI("http://example.root/item"), URIs.createURI("http://example.root/nested"), Kvin.DEFAULT_CONTEXT, 1619424246100L, 0, expectedValue));

		assertEquals(expected, tuples);
	}

	@Test
	public void shouldRejectMissingValueField() throws Exception {
		String json = "{ \"item1\": { \"prop1\": [ { \"seqNr\": 1 } ] } }";

		try {
			parse(json);
			fail("Expected RuntimeException due to missing 'value' field");
		} catch (RuntimeException e) {
			assertTrue(e.getCause() instanceof IOException);
		}
	}

	private List<KvinTuple> parse(JsonNode node) throws Exception {
		return parse(node, System.currentTimeMillis());
	}

	private List<KvinTuple> parse(JsonNode node, long currentTime) throws Exception {
		return parseBytes(mapper.writeValueAsBytes(node), currentTime);
	}

	private List<KvinTuple> parse(String json) throws Exception {
		return parse(json, System.currentTimeMillis());
	}

	private List<KvinTuple> parse(String json, long currentTime) throws Exception {
		return parseBytes(json.getBytes(StandardCharsets.UTF_8), currentTime);
	}

	private List<KvinTuple> parseBytes(byte[] bytes, long currentTime) throws Exception {
		JsonFormatParser parser = new JsonFormatParser(new ByteArrayInputStream(bytes));
		try (var tuples = parser.parse(currentTime)) {
			return tuples.toList();
		}
	}

	private JsonNode readResource(String path) throws Exception {
		InputStream stream = getClass().getResourceAsStream(path);
		try (stream) {
			assertNotNull(stream);
			return mapper.readTree(stream);
		}
	}
}